### DISCUSSION TITLE
Proposal: single-type files

### DISCUSSION THREAD

**Discussion by lostmsu** [dotnet/csharplang#928] [General]
This is related to "File scoped namespaces" proposal #137, I just wanted to track it separately (though IMHO, they both should be implemented in the same iteration).

Basically, the idea is to remove two indentation levels, that unnecessarily appear in C#<=7: namespace and class. Here's the example syntax I'd be looking for:

```c#
using System;
using System.Collections.Generic;

public class MyCompany.MyLib.MyClass<T> where T: IMyInterface<T>;

bool good = true;

public override string ToString() => $"is good: {this.good}";
```

We can decide if usings will be treated as being inside MyCompany.MyLib or outside. I'd prefer the first one, but the second option would be less of surprise for newcomers.

Same code without the feature for comparison:
```c#
using System;
using System.Collections.Generic;

namespace MyCompany.MyLib
{
    public class MyCompany.MyLib.MyClass<T> where T: IMyInterface<T>
    {
        bool good = true;

        public override string ToString() => $"is good: {this.good}";
    }
}
```

**Comment by iam3yal** [comment]
@lostmsu The other proposal is much more sensible in my opinion because it tries to reduce the nesting created by namespaces but this looks like a complete new language to me. 😆 

Edit: Rephrased.

**Comment by lostmsu** [comment]
@eyalsk same for me. However, I think its because we all so much used to the two levels of indentation, that their absence feels completely alien.
I can see this helping a lot in diff viewers though.

**Comment by iam3yal** [comment]
@lostmsu Can't see the benefit of this but each to his own I guess. :)

**Comment by lostmsu** [comment]
@eyalsk open this on laptop (random pull request from my browsing history): https://github.com/dotnet/roslyn/pull/20269/files
Note how many lines in C# code are wrapped in diffs because line is too long. This would reduce this effect. In general, it will be easier to program on laptop due to screen space restrictions.

**Comment by jnm2** [comment]
I am completely on board with including namespace and declaring type name inline with the class name, but I really am not liking the more drastic effect of making it look like the class declaration has ended but then including the following declarations inside it. I don't like it any more than I liked it in the `using (...);` statement proposal.

**Comment by lostmsu** [comment]
@jnm2 that makes sense. Unfortunately, some token is needed there to avoid ambiguity.

Perhaps it's better to replace `;` with `=>`. Although that would make indentation of `=>` inconsistent with indentation in functions.

Another alternative is to use some existing keyword like `is` in place of `;`.

But I am actually against that. Somehow this is similar to `using System;`. Despite `using` not being a declaration everyone knows its scope is till the end of file. And unlike `using` statement proposal you mentioned, in this case it is guaranteed to be the only statement like this in the file, and file will always start with it.

**Comment by jnm2** [comment]
> Despite using not being a declaration everyone knows its scope is till the end of file.

Same with variable declarations; and yet my point is, in neither case do these statements actually *contain* the following declarations or statements. With `using System;`, `System` does not contain what follows. With `int x;`, `x` does not contain the following lines. It's more like an attribute (in a limited way) in that it *applies* to the following declarations and statements, rather than *containing* them.

**Comment by MkazemAkhgary** [comment]
This is not same

```
namespace MyCompany.MyLib
{
    public class MyCompany.MyLib.MyClass<T> where T: IMyInterface<T>
```

I expect `MyCompany.MyLib.MyCompany.MyLib.MyClass`, why making c# syntax, a c++ like syntax? 😄 

Instead I would say namespaces could be omitted inside class name. so 
```
public class MyCompany.MyLib.MyClass<T> where T: IMyInterface<T>
{ // notice we didn't remove body.
}
```

would be equal to
```
namespace MyCompany.MyLib
{
    public class MyClass<T> where T: IMyInterface<T>
    {
```

IMO replacing body of class semicolon is not good suggestion. that's 👎  part but having namespace with class name for me is 👍 and doesn't make strange change in current syntax. (it saves 3 lines for good sake.)

**Comment by lostmsu** [comment]
@jnm2 you are appealing to consistency, but that consistency does not really exist, and two different kinds of `using` statement is a good demonstration of that. Note, that `using(x = new X())` has a body, while `using System;` just implicitly continues until the end of scope.

P.S. I intentionally did no say `using(var x = new X())`, because then you could argue, that `using(...)` defines something. Well. It does not! Actually, it just introduces a lifetime scope for an object. In that sense `using System;` also introduces visibility scope for namespace members.

**Comment by stepanbenes** [comment]
This syntax may collide with the proposed syntax for records.

**Comment by fanoI** [comment]
OK to remove namespace indentation but I prefer to leave the '{' for the class body, I'll interpret your example as:  ``MyCompany.MyLib.MyClass<T>`` is an empty class, ``bool good`` is a global and the rest are "free functions"...

This version of C# seems more a scripting language that a "true" programming language.

**Comment by jnm2** [comment]
@lostmsu I don't see any inconsistency. Technically there's one `using` statement; the namespace import thing is usually called a directive not a statement– but more importantly, the two things are not related to each other. They could have been called `foo` and `bar`.

Things that contain other things do it via blocks and indentation. Classes contain declarations, block statements contain statements, etc. There's a special exception for blocks containing one statement but no exception for blocks containing one declaration, but that's not what you're asking for either.

**Comment by ghost** [comment]
How would nested classes be handled? Nested classes are more common (and more useful) than nested namespace.

Nor am I fan of working in a codebase where it's impossible to add a new class to an existing file even on a temporary (i.e. pre-PR) basis. When I'm bringing up a big new feature with lots of new classes, it's just much more efficent to throw them into the file I'm working on during the bringup and segregate them in waves (during those otherwise unproductive 15min time-slices between meetings and a lunch or whatever) after the big thinking is done and I'm sure I actually want the classes.

**Reply by cbordeman** [reply]
Nested classes would require the old syntax.

**Comment by jnm2** [comment]
> Nor am I fan of working in a codebase where it's impossible to add a new class to an existing file even on a temporary (i.e. pre-PR) basis. 

^^ So much this.

**Comment by jnm2** [comment]
> How would nested classes be handled? 

I really (and I mean terribly) wish I could do this:

##### SomeClass.cs:
```cs
partial class Some.Namespace.SomeClass
{
    // ...
}
```


##### SomeClass.SomeNestedClass.cs:
```cs
class Some.Namespace.SomeClass.SomeNestedClass
{
    // ...
}
```

**Comment by lostmsu** [comment]
@jnm2 `using NS;`. Not sure what you mean by a directive. Just like `using(...){ }` it has scope of effect.
E.g.
```csharp
namespace A {
  using System;
  // can see Int32 here
}
namespace B {
  // using System;
  // can't see Int32 here
}
```

**Comment by jnm2** [comment]
> Not sure what you mean by a directive

They are not related to each other. One is a directive which changes the compile-time resolution of names, the other is a statement which has an effect at runtime.

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-directive
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-statement

**Comment by lostmsu** [comment]
@AtsushiKan I don't see a problem with nested classes. Since only topmost class can have this syntax, you can nest other classes inside without any ambiguity.
```csharp
class A.B.C;

enum D { Val }
```
other file
```csharp
... A.B.C.D.Val ...
```

**Comment by lostmsu** [comment]
@jnm2 I understand they are unrelated semantically. The point was that they both have scopes, and in one case scope is implicit, while in another it is explicit.
Is there a rule in language design, that would require code to be nested for one kind of scopes, and not nested for another?
F# seems to be doing fine without indentation for this
```F#
use x = new X()
x.bla()
```
and Java is ok for 
```java
package A.B.C;
```

**Comment by HaloFour** [comment]
@lostmsu 

It's not the `using` directive that has a scope, it's the `namespace` directive above it.

**Comment by lostmsu** [comment]
@HaloFour that's just a word play. Using directive has a scope, and that scope is the scope of namespace directive it is enclosed in.

**Comment by HaloFour** [comment]
@lostmsu 

It's not word play.  The scope is introduced by the `namespace` directive which uses block syntax, just like everything else in C#.  Eliminating that for types (`class`, `struct`, `enum`, etc.) I feel makes no sense within this language, or most languages, which use either some kind of block syntax or indentation to denote that relationship.

**Comment by lostmsu** [comment]
@HaloFour it is a word play, because I never said, that `using NS;` introduces scope. I just said it has one, and you argued assuming "has" == "introduces".

I do have the same feeling about block syntax for classes, but it is just because having explicit block and indentation is status quo, and we are used to it.

But rationally, there's a real problem of horizontal space inflation this restriction causes in over 95% of files, and the only reason not to have it fixed is that status quo (a better syntactical alternative would also do, but I can't come up with another one).

For another example of relatively successful language with similar approach, but different solution, take a look at Delphi:
```Delphi
unit A.B;
type C = class
  ...

implementation

procedure C.D();
begin
   ...
end;
```

Actually, even C++ does not have this problem for implementations due to header-implementation separation.

**Comment by HaloFour** [comment]
@lostmsu 

Sure `using` "has" a scope, just like virtually everything else in the language.  That scope is defined by the block syntax of the `namespace` directive.  You're the one that brought up `using` and how it somehow relates to your argument.  At best, it doesn't.

> But rationally, there's a real problem of horizontal space inflation this restriction causes in over 95% of files, and the only reason not to have it fixed is that status quo

This is the case in most languages.  You're free to at least not indent if you don't want to, unlike any of the whitespace sensitive languages which will force the issue.

> For another example of relatively successful language with similar approach, but different solution, take a look at Delphi

Delphi was based on Pascal and a completely different lineage from C.  Claiming that it's relatively successful is mildly amusing given that Delphi never broke meager marketshare even when it was at the top of its game.  Not that I have anything against Delphi.  That said, you are aware that Anders Hjelsberg designed both Delphi and C#, yes?

> Actually, even C++ does not have this problem for implementations due to header-implementation separation.

The ownership of the members is defined in the header which is where the scope blocks are required.  Like with C# the indentation is entirely optional.

I could buy the argument of eliminating the block for `namespace` directives.  That would kind of borrow a page from Java with its `package` directive and given that there are usually significantly fewer namespaces across a project vs. types I don't think it burdens the developer significantly.  But eliminating the block for types just makes the syntax look very messy.

**Comment by lostmsu** [comment]
@HaloFour  I don't want to argue the meta about if my `using` argument is relevant to discussion or not. If it does not for you - sure. You see it as completely irrelevant - sure, others might disagree.

These are the two statements, that I call "status quo":

> Delphi was based on Pascal and a completely different **lineage** from C

Now somehow lineage is supposed to define if a feature is useful and easy to understand?

> That would **kind of borrow a page from Java** with its package directive

So if Java would have that feature, you would be OK to _kind of borrow_ it? What if Java language authors though the same while designing language, and kept C++'s namespace block syntax just like C# did? You would not want to borrow namespace directive anymore?

P.S. my awareness about Andres Hjelsberg is completely irrelevant to this discussion. But you could guess since I gave an example in Deplhi.

**Comment by HaloFour** [comment]
@lostmsu 

> Now somehow lineage is supposed to define if a feature is useful and easy to understand?

No, but it sets an expectation as to how the syntax is laid out.  The designers can certainly deviate where they feel its appropriate.

> So if Java would have that feature, you would be ok to kind of borrow it? What if Java language authors though the same while designing language, and kept C++'s namespace block syntax just like C# did? You would not want to borrow namespace directive anymore?

Precedents help when determining whether or not a particular grammar is intuitive and makes sense.

In any case this is a purely stylistic proposal.  The maximum benefit it provides is the elimination of a handful of curly braces within the file.  The primary argument, "unnecessary indentation", can be solved today by the fact that C# doesn't care at all about indentation.  You are free to not indent the types defined within a namespace, and you are free to not indent the members defined within a type.

**Comment by lostmsu** [comment]
@HaloFour to some degree it is also partially about industry standard, and I would prefer to have less indentation to be widely adopted.

C# is on the front line. You can count languages, that have extraordinary features to borrow from, using fingers on just one hand.

**Comment by HaloFour** [comment]
@lostmsu 

Seems to be more the trend now to not only keep indentation but also to force it on the developer by making it semantically important.  You'd likely find more backers to a proposal for C# to eliminate curly braces in exchange for white-space sensitivity.

**Comment by lostmsu** [comment]
@HaloFour true, but this particular construct is special in that its usually singular in its file.

**Comment by HaloFour** [comment]
@lostmsu 

> but this particular construct is special in that its usually singular in its file.

Except when it's nested, which is far from uncommon in the C# world.

**Comment by DavidArno** [comment]
@AtsushiKan 
> How would nested classes be handled? 

To my mind, the clue to the answer to this is in the title: "Proposal: **single**-type files". If you want nested classes, use existing syntax.

> Nor am I fan of working in a codebase where it's impossible to add a new class to an existing file even on a temporary (i.e. pre-PR) basis.

Why would it be impossible? Just add `{}` to the class definition and now it supports multiple classes in the file.

**Comment by DavidArno** [comment]
For folk like me, who create a lot of small types, that I always want to be in their own files, this idea is a really nice one. Currently, especially for interfaces, I find the ceremony around defining the namespace and type can take up 50% or more of the file's content.

Being able to reduce, eg:
```cs
using System.Collections.Generic;

namespace SuccincT.Functional
{
    public interface IConsEnumerable<T> : IEnumerable<T>
    {
        ConsResult<T> Cons();

        IConsEnumerable<T> Cons(T head);
        IConsEnumerable<T> Cons(IEnumerable<T> head);
    }
}
```
to:
```cs
using System.Collections.Generic;

public interface SuccincT.Functional.IConsEnumerable<T> : IEnumerable<T> is

ConsResult<T> Cons();
IConsEnumerable<T> Cons(T head);
IConsEnumerable<T> Cons(IEnumerable<T> head);
```
is very appealing. All the noise and ceremony is stripped away, leaving a file half the size, which just contains information.

**Comment by HaloFour** [comment]
@DavidArno 

I've never really understood why C# developers so largely adopt the Allman style for braces while complaining about the extra newlines it requires.  I much prefer K&R (or at least the Java variant of it) and don't find your first example to have much in the way of "ceremony" as a result.  That's what I love about C#: all of that indentation and whitespace is completely optional.

I also prefer to have my top-level classes in individual files, including custom delegates.

**Comment by DavidArno** [comment]
@MkazemAkhgary,
> @DavidArno only up to this point is appealing. Removing body is meh

Then don't remove the body in your class declarations. 😕 

> ... Then you can have multiple types in one file...

The title of this suggestion is "Proposal: **single**-type files". I don't want to have multiple types in my files, ergo this feature is appealing to me. If you want multiple types in a file, then this feature is not for you.

@HaloFour 
> I've never really understood why C# developers so largely adopt the Allman style for braces while complaining about the extra newlines it requires. I much prefer K&R (or at least the Java variant of it)...

From experience, this is a non-sequitor argument as every Java developer I've met puts a blank line after the `{`, so it ends up the same number of lines:
```cs
using System.Collections.Generic;

namespace SuccincT.Functional {

    public interface IConsEnumerable<T> : IEnumerable<T> {

        ConsResult<T> Cons();

        IConsEnumerable<T> Cons(T head);
        IConsEnumerable<T> Cons(IEnumerable<T> head);
    }
}
```
Even with those extra lines removed, the indentation remains. Indentation after a `{` is so ingrained in the heads of millions of C-style language developers that laying the code out as the following would just cause mass confusion:
```cs
using System.Collections.Generic;

namespace SuccincT.Functional 
{
public interface IConsEnumerable<T> : IEnumerable<T> 
{
ConsResult<T> Cons();

IConsEnumerable<T> Cons(T head);
IConsEnumerable<T> Cons(IEnumerable<T> head);
}
}
```

The solution is to remove the need for the `{}` for a file containing a single type. It's simple, elegant and effective.

**Comment by HaloFour** [comment]
@DavidArno 

Hi.  I'm a Java developer.  I don't put newlines after closing braces.  Now you've met one.  😜 

My point is that you choose your verbosity.  The language only requires two characters to establish the scope within a type.  Newlines and indentation is all up to the developer.

You also demanding that F# nix the indentation that is required for everything?  Even without the braces I doubt too many developers would embrace the idea of not enclosing the members in some kind of scope denoted by indentation.

**Comment by DavidArno** [comment]
@HaloFour,
> Hi. I'm a Java developer. I don't put newlines after closing braces.

Um, I was talking about opening braces. I suspect this a simple typo on your part, but hey, if cheap points are offered... 😉 

> My point is that you choose your verbosity. The language only requires two characters to establish the scope within a type. Newlines and indentation is all up to the developer.

Technically, yes. But to mind at least, convention trumps personal preference a lot of the time. That's why, when writing JavaScript or Java, I follow the K&R brace layout. For C#, I use Allman. I prefer the latter, but I follow the conventions of the former two languages over my own preferences.

> You also demanding that F# nix the indentation that is required for everything?

F# is an odd one in this regard. Declare a module (which, as far as I know, is just a abstract final class in .NET, or "static" in C# terms) and no indentation is required for the functions therein. Declare an interface or class/struct though and indentation must follow that `=`. That inconsistency is one of the things I think F# got wrong.

**Comment by HaloFour** [comment]
@DavidArno 

A newline *after* the opening brace?  I've never seen that.  Java devs seem to prefer fairly dense code, to the point of having closing brace, keyword and opening brace all on the same line, e.g.:

```java
if (foo) {
    doFoo();
} else {
    doBar();
}
```

**Comment by lostmsu** [comment]
@HaloFour newline after opening brace is default in IDEA for method bodies AFAIK. At least I've seen it a lot. Not for IFs.

I also prefer opening brace on the same line btw. Saves vertical space.

I am sad to see people downvoting feature that brings physically measurable advantage over their idea of language consistency. I fail to see how anyone even with least experience and involvement could not comprehend and remember that `class ...;` does not indicate empty body, but a "directive".

**Comment by iam3yal** [comment]
@lostmsu 

> I am sad to see people downvoting feature that brings physically measurable advantage over their idea of language consistency. 

C# is a curly brace (or bracket) language not because the "userbase" chose it to be but because the designers of the language chose it and we followed partly because the syntax is similar to what we are accustomed to, most notably the curly braces so the idea wasn't ours to begin with.

You see that as a benefit some of us see that as a fundamental change to the language that we're not in favor of or think that it's too extreme and this is the reason for the downvotes.

p.s. I really wouldn't mind it either way so I removed the downvote, I still think it's a fundamental change to the language and one that I probably wouldn't use but don't care whether others will.

**Comment by sharwell** [comment]
💭 Java approaches this functionality in a slightly different manner:

* The `package` clause is written in "statement" form instead of "block" form
* Various IDEs expose an option to not require indentation for top-level members of a type

If this approach (both options) were applied to C#, the resulting code might look like this:

```
using System;
using System.Collections.Generic;

namespace MyCompany.MyLib;

public class MyClass<T> where T: IMyInterface<T>
{
bool good = true;

public override string ToString() => $"is good: {this.good}";
}
```

💭 My biggest concern overall with this proposal is the confusion and/or limitations that may result from the placement of `using` directives. In the current C# language, `using` directives have slightly different semantics when they are placed inside a `namespace` declaration instead of outside it.

💭 I also have a concern related to the `class Namespace.Type { }` form:

* If you only allow the dotted names for non-nested types, then some benefits of single-file types will not be achievable for nested types in a `partial` class
* If you only allow the dotted names for nesting types, then you would not be able to remove the separate `namespace` declaration line
* If you try to allow something like `partial class Namespace.Outer.Nested { }`, then the behavior of the single-file type becomes dependent on other files (is `Outer` a namespace or a type? what about `Namespace`?).

**Comment by jnm2** [comment]
> If you try to allow something like partial class Namespace.Outer.Nested { }, then the behavior of the single-file type becomes dependent on other files (is Outer a namespace or a type? what about Namespace?

Absolutely it's dependent, and I like it that way. Any time you have `partial` in the class definition, you know you are depending on other files in the first place.

**Comment by AustinBryan** [comment]
I think it's a good idea to remove the namespace indents, but I would feel like everything I'm writing is in some global scope without the class. It feels like I'm writing in C or something 😐

**Reply by GeraudFabien** [reply]
It would feel like this at the beginning. But has everythinks in any language you have to take time to adapt. And i don't thinks it would be long.

It would help mitigate some issue : 
- When you open a a source file on half or else of you're screen it's really hard to read since all line are by default indent
- Enforce the fact the file should only contain one class (Even if we can support nested class we can't support two class not nested with this system). (Personnaly I'm on the side not to support nested because it would simplify avoid a lot of problem)
- When you have a big file sometime it's hard to keep track of closing bracket. Without a level a not needed closing bracket will directly show the right error near the right place.
- When a dev is too lazy and do multiline string it would remove a lot of useless space in the final string.

The main problem for me is that it must be forbidden to write the file scope class sentence affter the first method/Property/field/... is specify.

**Comment by DavidArno** [comment]
It's interesting revisiting ideas like this years later. 

Back then, I was really keen on that idea. Since then, the language has added support for file-scoped namespaces. But it has also added top-level statements. To my mind, that latter feature changes this one from one I was keen on to one I'm strongly opposed to:
```c#
// Not a class, just some top level statements
namespace MyNamespace;

string field = 1;

void Bar() => field = 2;

string Foo() => field;


// Is a class
namespace MyNamespace;
class SomeClass;

string field = 1;

public string Foo() => field;

void Bar() => field = 2;
```
The only clues that one is a class and one isn't is that `class` line and that `Foo` is marked as `public` in one and not the other. When a file is possibly hundreds of lines long, that sort of thing is too easily missed.

**Reply by lostmsu** [reply]
It might be better to have `class MyNamespace.SomeClass` without the namespace directive.

Besides, I think top-level members are only allowed in the file that is the program entry point, which is usually hard to miss.

**Reply by CyrusNajmabadi** [reply]
> It might be better to have class MyNamespace.SomeClass without the namespace directive.

This woudl be highly ambiguous with other proposals around allowing that for nested types.  Paging @jnm2

**Reply by scottmwyant** [reply]
I like the proposal and think it aligns nicely with file level namespaces.  The sample you shared is nasty but I think you’d limit it to either top level (which is implied `class Program`) or a single class declaration ending with `;`. Nested classes would use conventional braces.

**Reply by jnm2** [reply]
Yes, `class A.B.C { }` would have to mean `class C { }` nested inside partial type `B` nested inside partial type `A`. I think @333fred or others were commenting in Discord a long while ago on how it would be unworkable for these to be namespaces.

I do have a use case for `class Outer.Nested { }` related to what happens to source control line history and lack of rename history if you refactor a nested class to a top-level one or vice versa. I'm planning on drafting a proposal.

**Reply by SpocWeb** [reply]
I would rather swap Top-level statements with file-scoped classes, 
because for each Program you write, you also write hundreds (or thousands) of classes.

**Comment by 333fred** [comment]
See also the answers for https://github.com/dotnet/csharplang/discussions/4722, where Cyrus and I laid out why this won't happen.

**Reply by lostmsu** [reply]
The syntax is just an example. Semicolon can be replaced by a keyword. `is`, `with`, or any other token really.

**Reply by 333fred** [reply]
In general, we're not trying to remove the braces here. We find it visually ambiguous and not worth the effort, regardless of what token ends the `class` line.

**Reply by CyrusNajmabadi** [reply]
> The syntax is just an example. Semicolon can be replaced by a keyword. is, with, or any other token really.

Can you show examples of what you mean?  As this proposal is entirely about syntax, we would need to see what form you mean here.  Thanks!

**Reply by lostmsu** [reply]
I mean just take the original suggestion and replace semicolon after the class title with anything. E.g.

```csharp
using System;
using System.Collections.Generic;

public class MyCompany.MyLib.MyClass<T> where T: IMyInterface<T> with

bool good = true;

public override string ToString() => $"is good: {this.good}";
```

**Comment by Mindstyler** [comment]
So C# recently introduced a `file` keyword for file scoped methods, etc ...
And I was thinking maybe this could also be used to remove ambiguity in this use case by being added as another explicit accessibility modifier.
AKA something like
```cs
public file class MyClass;
```

Looking forward to your thoughts.

**Reply by 333fred** [reply]
The file keyword is specifically for making a type visible only in the current file. It cannot be used for this purpose.

**Reply by Mindstyler** [reply]
Well yeah, but it's use could be extended so it could be used here.

**Reply by 333fred** [reply]
No it couldn't. It already means something when applied to types. That something is not this.

https://github.com/dotnet/csharplang/discussions/928#discussioncomment-1877393 as well, where I pointed out that the LDM has explicitly rejected this concept.

**Reply by Mindstyler** [reply]
Ah damn. I wasn't aware it's actually already a valid modifier for types.
I don't mind writing the normal classes, just thought that this might've worked.

**Reply by calin-darie** [reply]
`file` may not be an option, but there could be another keyword, right?

**Reply by calin-darie** [reply]
Also maybe we should force the keyword indicating file scope as the first keyword, so that it stands out more

**Comment by pelletburner** [comment]
I agree with this proposal, but the ";" at the end of the class declaration needs to go and be replaced with something else.

Although I understand that this is there to put it in line with the `namespace` declaration syntax.

**Reply by Eli-Black-Work** [reply]
A colon (`:`) seems more intuitive, to me 🙂

**Reply by pelletburner** [reply]
> A colon (`:`) seems more intuitive, to me 🙂

Yes, was thinking the same.

";" is so "final", to me this implicates that "something is done".

**Reply by UniqeId6542** [reply]
I want to agree, but...

`namespace MyNamespace;` has the rest of the file under the namespace
`using System;` has the namespace available for the rest of the block / file
`using (MyDisposable d = new());` has d available for rest of the block and dispose it after
`int x;` has x available for the rest of the block

In my head, I want a statement to be more 'isolated' (`a = b + c;`)  But maybe my concept of a statement needs broadened.

**Comment by InCerryGit** [comment]
It looks great, it can save screen space. Is there any progress on this?

**Reply by DavidArno** [reply]
I'm not sure why this discussion wasn't closed as members of the language team made clear a while ago that this will never happen. See #4722 for all the details, but effectively it is incompatible with records and primary constructors as they already support declaring a type terminating the definition with `;`

```cs
// This could work:
public class MyClass;

bool good = true;

public override string ToString() => $"is good: {this.good}";

// This wouldn't:
public class MyClass(int x); // This is a complete primary constructor declaration so the rest of the file would be an error 

bool good = true;

public override string ToString() => $"is good: {this.good}";
```

**Reply by saint4eva** [reply]
Kindly get bigger monitors if you need more screen space.

**Reply by Eli-Black-Work** [reply]
@DavidArno Would `public class MyClass:` work? (Colon instead of semicolon)

**Reply by InCerryGit** [reply]
> Kindly get bigger monitors if you need more screen space.
 
Many people work with laptops, so the screen size is limited.

**Reply by DavidArno** [reply]
> @DavidArno Would `public class MyClass:` work? (Colon instead of semicolon)

It *could*, but as I said before, the language team has made it quite clear they aren't interested in this feature, eg see [here](https://github.com/dotnet/csharplang/discussions/928#discussioncomment-1877427) and [here](https://github.com/dotnet/csharplang/discussions/928#discussioncomment-3543460). It's unclear to me why they haven't just posted a "We won't ever do this" comment and then marked the discussion as answered as it is misleading as it stands and appears the topic is still up for debate.

**Reply by Eli-Black-Work** [reply]
@DavidArno Thanks, I missed that the first time around 🙂

I admit, I do like that the C# language designers leave discussions like this open, rather than locking them, so that there's an opportunity for continued discussion.

**Comment by DeminAO** [comment]
it might look like this:
- the definition of the structure (class, interface, etc.) has no scope - the entire file is the scope of the structure
- the directory path of the structure is the definition of their namespace
- by default, it would be possible to omit the class definition and consider the file name as its name (it would be possible to define the type of structure (class, enum, interface, record) in the file name)
- for the simplicity of defining the constructor, one could use ".ctor". Then there would be no need to worry about updating the file name and constructor definition inside
- or completely abandon constructors in favor of field attributes, if we are talking about injecting services from a container. For example, so: `inject ISomeService someService`

**Reply by DeminAO** [reply]
and then, using globalusings, omitting the namespace definition, removing the class brackets, removing the constructor, we can get clean code in which only logic is concentrated.

example:
.proj
-- PeopleModule
-- -- ContextRepository.cs
-- -- Domain
-- -- -- PeopleDbContext.cs
-- -- -- Entities
-- -- -- -- Person.cs

```csharp

// PeopleDbContext.cs

implements DbContext;

public DbSet<Entities.Person> People { get; set; }

protected override void OnModelCreating(ModelBuilder modelBuilder)
{
	// configure people
}
```

```csharp

// ContextRepository.cs

public interface IContextRepository
{
	Person GetPerson(int id);
}

implements IContextRepository;
inject Domain.PeopleDbContext _context;

public Person GetPerson(int id)
{
	return _context.People.Find(id);
}
```

**Reply by DeminAO** [reply]
and at controller:

.proj
-- PeopleModule
-- -- PeopleController.cs

```csharp
implements ControllerBase;

inject ContextRepository.IContextRepository _repository;

[HttpGet]
public ActionResult<Person> GetPerson([FromQuery] int id)
{
	return Ok(_repository.GetPerson(id));
}
```

**Reply by DeminAO** [reply]
about constructors

public PeopleDbContext(DbContextOptions<PeopleDbContext> options) : base(options) { }

overriding a constructor in a child class could be hidden from the developer's eyes and implemented if the developer did not add it. That is, the entry above could be omitted

**Comment by AndreLeria** [comment]
I don't think anyone suggested this yet: why don't we couple the namespace declaration with the class declaration?

```csharp
namespace MyNamespace.MySubNamespace with
class MyClass : IWhatever;
```

In which case the `namespace (...) with` would mean this is a single-type file. Then add a newline before or after the `with` keyword when formatting.

**Reply by Mike-Logit** [reply]
Despite the controversy around this feature, I agree that `with` makes the most semantic sense and it also fixes the issue of visually identifying what's going on from a code review perspective. If this feature were to be added, I'd vote for this approach.

**Reply by lostmsu** [reply]
I kinda like the DeminAO's approach where the type is implicitly defined based on the file name.
